sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -361,6 +361,15 @@ class JacksonParser(
         // For such records, all fields other than the field configured by
         // `columnNameOfCorruptRecord` are set to `null`.
         throw BadRecordException(() => recordLiteral(record), () => None, e)
+      case e: CharConversionException if options.encoding.isEmpty =>
+        val msg =
+          """Failed to parse a character. Encoding was detected automatically.
```

Comment:
I think automatic detection is true only when multuline is enabled. We can just describe it in documentation and, reward this message or even just remove this message too.

Reply:
This message was added explicitly to tell our customers how to resolve issues like https://issues.apache.org/jira/browse/SPARK-23094 .  Describing that in docs is not enough from our experience. Customers will just create support tickets, and we will have to spend time to figure out the root causes. The tip can help the customers to solve the problem on their side. /cc @brkyvz

Reply:
My point is automatic detection is true only when multuline is enabled and the message looks like it's always true. I don't think we should expose an ancomplete (or accidential) functionality in any case.

Reply:
ok, speaking about this concrete exception handling. The exception with the message is thrown ONLY when options.encoding.isEmpty is `true`. It means `encoding` is not set and actual encoding of a file was autodetected. The `msg` says about that actually:  `Encoding was detected automatically`.

Maybe `encoding` was detected correctly but the file contains a wrong char. In that case, the first sentence says this `Failed to parse a character`. The same could happen if you set `encoding` explicitly because you cannot guarantee that inputs are alway correct.

> I think automatic detection is true only when multuline is enabled.

Wrong char in input file can be in a file with UTF-8 read with `multiline = false` and in a file in UTF-16LE with `multiline = true`.

My point is the mention of the `multiline` option in the error message doesn't help to user to solve the issue. A possible solution is to set `encoding` explicitly - what the message says actually.

Reply:
I don't think `Encoding was detected automatically` is not quite correct. It might not help user solve the issue but it gives less correct information. They could thought it detects encoding correctly regardless of `multiline` option.

Think about this scenario: users somehow get this exception and read  `Failed to parse a character. Encoding was detected automatically.`. What would they think? I would think somehow the file is somehow failed to read but it looks detecting the encoding in the file correctly automatically regardless of other options.

It's annoying to debug encoding related stuff in my experience. It would be nicer if we give the correct information as much as we can.

Reply:
I am saying let's document the automatic encoding detection feature only for `multiLine` officially, which is true.

Reply:
> I don't think `Encoding was detected automatically` is not quite correct.

It is absolutely correct. If `encoding` is not set, it is detected automatically by jackson.  Look at the condition `if options.encoding.isEmpty =>`. 

> It might not help user solve the issue but it gives less correct information.

It gives absolutely correct information.

> They could thought it detects encoding correctly regardless of multiline option.

The message DOESN'T say that `encoding` detected correctly.

> Think about this scenario: users somehow get this exception and read Failed to parse a character. Encoding was detected automatically.. What would they think?

They will look at the proposed solution `You might want to set it explicitly via the encoding option like` and will set `encoding`

> I would think somehow the file is somehow failed to read

It could be true even `encoding` is set correctly

> but it looks detecting the encoding in the file correctly automatically 

I don't know why you decided that. I see nothing about `encoding` correctness in the message.

> It's annoying to debug encoding related stuff in my experience. It would be nicer if we give the correct information as much as we can.

What is your suggestion for the error message?

> I am saying let's document the automatic encoding detection feature only for multiLine officially, which is true.

I agree let's document that thought it is not related to this PR. This PR doesn't change behavior of encoding auto detection. And it must not change the behavior from my point of view. If you want to restrict the encoding auto-detection mechanism somehow, please, create separate PR. We will discuss separately what kind of customer's apps it will break.